Remove network calls from rendering `App` in tests

dquenne opened the pull request:
Drop default create-react-app test on `App.tsx` and add a couple simple tests to `Birthday.tsx` that use a stub provider for `uqrl` to avoid network calls from the real web client-based provider in `App.tsx`. Also, add a fallback mock of `window.fetch` to ensure that no network calls ever slip through from the test runner.

@rstoutamore this may conflict with the create-react-app to vite migration, so let me know if you'd like me to hold off on merging this.

rstoutamore committed:
Remove network calls from rendering `App` in tests

Drop default create-react-app test on `App.tsx` and add a couple simple
tests to `Birthday.tsx` that use a stub provider for `uqrl` to avoid
network calls from the real web client-based provider in `App.tsx`.
Also, add a fallback mock of `window.fetch` to ensure that no network
calls ever slip through from the test runner.

Co-authored-by: Dylan Quenneville <12236876+dquenne@users.noreply.github.com>

rstoutamore reviewed (approved):
Your changes looked good to me @dquenne - could you take a peak and sanity check that I didn't break anything when rebasing? Otherwise good to go

rstoutamore commented:
Oh hm wait a sec, I accidentally left `App.test.tsx` in when merging, but that test is still passing despite attempting to fetch.

dquenne commented:
If it's helpful, App.test.tsx definitely was failing on Jest, so I suspect it's some difference with how the error bubbles up between the two test runners.

rstoutamore committed:
Adds an assertion that mock fetch was not called

I don't like doing this in a teardown function, but I think we lack good
alternatives given react-query's handling of any errors thrown while
attempting to fetch. I left the mock behavior in place, because it could
catch other usage of fetch, and the message is still present in the
output and clear. I couldn't present any more descriptive an error via
Jest without janky workarounds or adding a new package.

App.test.tsx fails at this commit to prove that the test works. Removing
in next commit, and will squash and merge.

rstoutamore committed:
Delete `App.test.tsx`

rstoutamore commented:
Turns out that `App.test.tsx` passes because react query was just handling any errors, including the error raised by the mock implementation of `fetch`. Added a check that fetch was not called and validated that `App.test.tsx` subsequently failed, more detail in the commit message of https://github.com/dquenne/anime-overlap/pull/17/commits/e67af72b43b8ec63ea9d7c26ec57c5fb51a20009. Don't love the solution, but couldn't get anything to work from within the mock implementation given react query's behavior.

rstoutamore commented:
> Turns out that `App.test.tsx` passes because react query was just handling any errors, including the error raised by the mock implementation of `fetch`. Added a check that fetch was not called and validated that `App.test.tsx` subsequently failed, more detail in the commit message of [e67af72](https://github.com/dquenne/anime-overlap/commit/e67af72b43b8ec63ea9d7c26ec57c5fb51a20009). Don't love the solution, but couldn't get anything to work from within the mock implementation given react query's behavior.

Never mind, after thinking on this some more, I do not like this solution anymore. Assertions in teardown seemed like a bad idea anyways, but I think it can also just lead to incorrect behavior. Will poke at this next week

rstoutamore committed:
Revert "Adds an assertion that mock fetch was not called"

This reverts commit e67af72b43b8ec63ea9d7c26ec57c5fb51a20009.
